---
author: "Ewy Mathe, Elizabeth Baskin and Rick Farouni"
title: "ALTRE: vignette"
date: "`r Sys.Date()`"
output: pdf_document
vignette: >
  %\VignetteIndexEntry{ALTRE: vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


##Introduction

Regulatory elements (REs) (e.g. promoters, enhancers) modulate transcription and affect cellular development using a variety of mechanism: DNA looping, the production of enhancer RNAs (eRNAs), and enhancer-promoter interactions among them. Identifying altered REs between cell lines and tissue types holds promise for a variety of applications, including identifying new mechanisms involved in disease progression, investigating how subtypes of broader diseases differ from one another, or focusing on those REs that do not change (to better understand the housekeeping functions of cells). 

Assays such as ATAC-seq, DHS-seq and FAIRE-seq can identify RE activity by identifying open, transcriptionally active chromatin. These assays produce files of known peak locations (called hotspot files). The hotspots are regions of open chromatin marking promoters and enhancers. While these assays are more and more frequently performed in cell lines and clinical samples, there are currently few resources available to guide the raw data to meaningful results. This package is the first to bundle the various scripts and steps of RE analysis into a single, compact tool that guides users through a defined workflow. Previously, individual data analyses pipelines were strung together as needs arose, making reproducible, shareable data-analysis a challenge -- an especially difficult challenge for newcomers to the field, especially those with little programming skill. 

Additionally, ALTRE defines altered peaks using both binary data (presence/absence of peaks) and quantitative data (peak intensity), includes functions that merge and annotate peaks (e.g. as promoters and enhancers), performs enrichment analysis for REs of interest, provides visualization functions, and includes a web interface using Rshiny for those not as fluent with the R statistical language.  

This vignette further explains the purpose of the package and establishes the workflow. 

##Workflow

#Before beginning 

#Input data 

ATAC-seq, DHS-seq and FAIRE-seq are high-throughput sequencing techniques for identifying regions of open, transcriptionally active chromatin. As a result, short sequencing fragments will reflect open chromatin, enabling researchers to easily distinguish between open and closed chromatin. Open chromatin will generate peaks for analysis, closed chromatin will not. While ATAC-seq, DHS-seq and FAIRE-seq share many similarities, ATAC-seq has gained momentum recently, as it requires significantly fewer cells, takes less time, and is less labor-intensive.  

As for all high-throughput sequencing experiments, the results of ATAC-seq, DHS-seq and FAIRE-seq experiments are delivered in FASTQ files. The initial processing of the FASTQ file must be accomplished outside of ALTRE, as the R-package requires BAM and hotspot files as inputs. Alignment of sequencing reads produces BAM files, and peak identification produces hotspot files. There are a number of software packages available for both alignment and hotspot calling; we recommend bowtie for alignment, and dnase2hotspot for hotspot calling.


#Example data 

To demonstrate the abilities of ALTRE, we have provided an example dataset (BAM and HOTSPOT files) of cancerous (A549) and normal (SAEC) lung cell types downloaded from the ENCODE project. This dataset will guide users through the workflow of the package in this vignette. 

This vignette uses only the chr21 subset of the data. This data is available here: https://mathelab.github.io/ALTREsampledata/, and we will automatically download it below for this vignette. 

##Start of pipeline

#Data Preparation:
#Reproducible peaks

Before analysis, it is important to identify consensus peaks amoung sample replicates -- peaks that are present in all (or, less stringently, the majority) of sample replicates. The function "getConsensusPeaks" processes an input of two or more hotspot files per sample and returns a GRanges list. At least two hotspot replicates are required in order to differentiate sample noise from robust, reproducible peaks. Hotspot files must be formatted as "chr", "start", "stop" in the first three columns. Additional columns are allowed, but will be ignored. 

Many of the functions within ALTRE rely on meta-information about the samples supplied in a csv file. 

First, we need to download all the necessary sample files for this analysis and then read in a csv file that contains meta-information about our samples. This csv file is included in the bundle of chr21 subsetted data. 
```{r}
library(ALTRE)
download.file("https://github.com/Mathelab/AltreDataRepo/tarball/master", "./dataset.tar.gz")
untar("dataset.tar.gz")
workingdirectory=getwd()
csvfile <- file.path(workingdirectory, "Mathelab-ALTREsampledata-042a6f3/DNaseEncodeWindows.csv")
csvfile
```

In this vignette we will change information in the csv file on the command line, by editing the object. It likely makes more sense for you to manually edit the file when running 
this analysis on your own samples. 
```{r}
sampleinfo <- loadCSVFile(csvfile)
filespath <- file.path(workingdirectory, "Mathelab-ALTREsampledata-042a6f3")
sampleinfo$datapath=filespath
```

For the "getConsensusPeaks" function, we need to read in the hotspot files for analysis. 
```{r}
samplePeaks <- loadBedFiles(sampleinfo)
samplePeaks
```

Finally, we use the "getConsensusPeaks" function to identify the peaks that are present in both replicates of each sample type. 
```{r}
consPeaks <- getConsensusPeaks(samplepeaks = samplePeaks, minreps = 2)
```

"plotConsensusPeaks" creates a bar graph comparing the number of replicate peaks identified vs. total number of peaks.
```{r}
plotConsensusPeaks(samplepeaks = consPeaks)
```

#RE annotation 

The output of “getConsensusPeaks” is then further processed by the “combineAnnotatePeaks” function, which accomplishes three main tasks:

1. Combines the reproducible hotspots from both samples type into one large master list of hotspots. This enables peak height comparison across sample types.

2. Categorize the hotspots as either promoters or enhancers based on the distance of the RE from a transcription start site (TSS). In this vignette we will use the default promoter distance argument: REs within 1,500 bp of a TSS are considered promoters; those not within 1,500 bp of a TSS are considered enhancers.

3. Optionally, merges REs within a certain distance of each other. Merging close-by REs loosens the stringency of the region comparison. Multiple regions of active chromatin within ~1000 bp are likely to represent only one RE. In this vignette we will use the defaults of the function, which merges regions within 1000 bp of each other and merges only within RE type  (i.e. only enhancers will only be merged with other enhancers, and likewise, promoters only with promoters). Both these functionalities can be changed via the "mergedistenh", "mergedistprom" and "regionspecific" arguments. 

This function requires a list of TSSs for the promoter/enhancer annotation: 
```{r}
TSSannot <- getTSS()
```

The output of the previous function, "getConsensusPeaks", is fed into this function:
```{r}
consPeaksAnnotated <- combineAnnotatePeaks(conspeaks = consPeaks,
                                           TSS = TSSannot,
                                           merge = TRUE,
                                           regionspecific = TRUE,
                                           mergedistenh = 1500,
                                           mergedistprom = 1000)
```


The function produces a GRanges object containing the annotated, merged hotspots -- all hotspots present in cancerous or normal lung. 

Additionally, the function "plotCombineAnnotatePeaks" creates a bar graph to showcase the changes in the number and size of enhancers and promoters before and after merging close-by REs. 

```{r}
plotCombineAnnotatePeaks(consPeaksAnnotated)
```

The number of REs decreases and the size of the regions increases, as expected.

##Quantifying RE acessibility

The next function, "getcounts" counts the number of reads in each RE for each sample type – this determines the peak height/intensity and approximates the accessibility of the RE in question.   

Read count information across regions of interest is only available in the BAM file format – BAM files must be supplied to this function for each sample type. At least two replicates per sample type are required in order to distinguish reproducible peaks from sample noise. The names of the BAM files are not supplied to the function directly. Instead, the information is included in the csv file containing sample meta-information. We have already extracted that information from the csv file in the sampleinfo object.


One cell type must be designated as the "reference type", to which the other cell types will be compared. In this example, the reference type is the normal lung cell line (SAEC). 
```{r}
counts_consPeaks <- getCounts(annotpeaks = consPeaksAnnotated,
                              sampleinfo = sampleinfo,
                              reference = 'SAEC',
                              chrom = 'chr21')
plotgetcounts(counts_consPeaks)
```




##Identification of cell-type specific REs

The count information from "getcounts" is then processed by "countanalysis" -- this is the point in the pathway where the changes in peak hieght across cell types are quantified with an algorithm from the DESeq2 R package.

```{r}
altre_peaks <- countanalysis(counts = counts_consPeaks,
                             pval = 0.01,
                             lfcvalue = 1)

```

Finally, altered enhancers and promoters are identified. Larger log2fold changes and smaller p-values signify greater differences between cell types. How large a change is considered cell-type specific is left up to the user. 

```{r}
categaltre_peaks <- categAltrePeaks(altre_peaks,
                                    lfctypespecific = 1.5,
                                    lfcshared = 1.2,
                                    pvaltypespecific = 0.01,
                                    pvalshared = 0.05)
plotCountAnalysis(categaltre_peaks)
```

REs with large, positive log2fold changes and small p-values are more open in the lung cancer cell type, and therefore, their activity is associated with the cancer phenotype. REs with large, negative log2fold changes and small p-values are more open in the normal lung cell type, and may therefore be associated with tumor suppressive properties. REs with minimal log2fold change and high p-values are equally open in both cell types, and therefore, their activity is more likely to contribute to the house-keeping functions required in every cell. 

#Pathway enrichment 

Finally, the results of the “categAltrePeak” can be used to identify whether there is an enrichment of certain pathways in the type-specific or shared regions using the “enrichment” function. This is accomplished by linking each RE to the closest gene, and then linking each gene’s product to the pathway the gene product participates in. Pathways that recur many times in the gene cluster (“enriched” pathways) are likely to be regulated by the REs linked to the gene cluster (whether type-specific or shared).

Pathways that are unlikely to be of interest can be filtered out of the enrichment results. ALTRE has two metrics to identify "low information" pathways.  

1. Vague pathways such as “DNA binding” are removed -- they encompass a number of other, more precise pathways (termed “offspring” pathways) already present in the enrichment results.
The "offspring" argument limits how many offspring a pathway can contain. Generally, the more offspring, the vaguer and less informative the pathway. 

2. Removal or pathways with low gene counts – these pathways are more likely to be false positives. This is because they are more likely to be elevated in the pathway rankings due to the nature of the statistical test – it is much easier to contain half of a pathway’s genes in the gene cluster under analysis when the pathway contains only four genes, and not 100 genes. The "gene" argument limits how few genes a pathway can contain. 

Pathways fall into three categories (Molecular Function (MF), Biological Process (BP), and Cellular Component (CC)), and each category must be analyzed separately. The category is selected by supplying an argument to "ontoltype". No pathways will have low p-values (p<0.05) since we are using only a chr21 subset of the data in this vignette  -- the p-value argument is set very high so that the results do not come up as "none" in this example. During a real analysis the p-value should be set to a more scientifically meaningful p-value cut-off, such as 0.05. The enriched pathways for this entire dataset at a low p-value cut-off can be seen in the published paper. 
```{r}
MFenrich <- pathenrich(analysisresults = categaltre_peaks,
                          ontoltype = 'MF',
                         enrichpvalfilt = 0.99)

```

##Post-processing summarization and visualization 

Once results have been obtained, there are a number of additional functions in ALTRE to summarize and better depict the results and raw data in tables and graphs. In this section of the vignette we will walk through these functions.

“comparePeaksAltre” creates a table to compare the two methods of identifying altered REs – one based on peak intensity, the other on peak presence or absence as determined by hotspot calling algorithms. The intensity-based method identifies much fewer type-specific regions. 

```{r}
analysisresults <- comparePeaksAltre(categaltre_peaks, reference= "SAEC")
```

"plotallvenn" takes the results from "comparePeaksAltre" and translates the table into venn diagrams.

```{r}
plotallvenn(analysisresults)
```

"plotDistCountAnalysis" enables users to view the raw count data in regions identified as type-specific or shared. The log2 transformation of reads per kilobase of RE per million is plotted. As expected, for REs identified as higher log-fold change than normal, the lung cancer cell line has much higher counts than he normal lung cell line. The opposite is true for the lower log-fold change regions. For RE with little log-fold change between cell types, the counts are similar. This visual depiction confirms the results of "countanalysis" and enables users to change parameters if needed. 

```{r}
plotDistCountAnalysis(categaltre_peaks, counts_consPeaks)

```

"writeBedFile" allows raw data to be visualized as tracks in the UCSC by creating hotspot files color-coded by type-specificity. A bed file will be created in the working directory, there is no output to the function that can be saved as an object.

```{r}
writeBedFile(categaltre_peaks, "output.txt")
```


"enrichHeatmap” creates a heatmap of the analysis for all three RE types (SAEC-specific, A549-specific, and shared), with the color-coding corresponding to the significance of the pathway – the lighter the blue, the lower the p-value. No pathways will have low p-values (p<0.05) since we are using only a chr21 subset of the data in this vignette  -- the p-value argument is set very so that the results do not come up as "none" in this example. During a real analysis the p-value should be set to a more scientifically meaniningful p-value cut-off, such as 0.05. The enriched pathways for this entire dataset at a low p-value cut-off can be seen in the published paper. 

```{r}
enrichHeatmap(MFenrich, title="GO:MF", pvalfilt = 0.99)

```

Pathways that are identified as enriched in all three RE types can be filtered out at this step.
